checkstyle-tester/diff.groovy

```diff
@@ -0,0 +1,243 @@
+        deleteDir(reportsDir)
+        moveDir(tmpReportsDir, reportsDir)
+        generateDiffReport(reportsDir, masterReportsDir, patchReportsDir, checkstyleCfg)
+        generateSummaryIndexHtml(diffDir)
+    }
+    else {
+        throw new IllegalArgumentException('Error: invalid command line arguments!')
+    }
+}
+
+def areValidCliArgs(options) {
+    def valid = true
+    if (options == null) {
+        valid = false
+    }
+    else {
+        if (options.checkstyleCfg
+                && options.localGitRepo
+                && options.patchBranch
+                && options.projectsToTestOn) {
+            def localGitRepo = new File(options.localGitRepo)
+            def patchBranch = options.patchBranch
+            def baseBranch = options.baseBranch
+            if (!isValidGitRepo(localGitRepo)
+                    || !isExistingGitBranch(localGitRepo, patchBranch)
+                    || (baseBranch && !isExistingGitBranch(localGitRepo, baseBranch))) {
+                valid = false
+            }
+        }
+        else {
+            err.println 'Error: wrong number of command line arguments!'
+            valid = false
+        }
+    }
+    return valid
+}
+
+def isValidGitRepo(gitRepoDir) {
+    def valid = true
+    if (gitRepoDir.exists() && gitRepoDir.isDirectory()) {
+        def gitStatusCmd = "git status".execute(null, gitRepoDir)
+        gitStatusCmd.waitFor()
+        if (gitStatusCmd.exitValue() != 0) {
+            err.println "Error: \'${gitRepoDir.getPath()}\' is not a git repository!"
+            valid = false
+        }
+    }
+    else {
+        err.println "Error: \'${gitRepoDir.getPath()}\' does not exist or it is not a directory!"
+        valid = false
+    }
+    return valid
+}
+
+def isExistingGitBranch(gitRepo, branchName) {
+    def exist = true
+    def gitRevParseCmd = "git rev-parse --verify $branchName".execute(null, gitRepo)
+    gitRevParseCmd.waitFor()
+    if (gitRevParseCmd.exitValue() != 0) {
+        err.println "Error: git repository ${gitRepo.getPath()} does not have a branch with name \'$branchName\'!"
+        exist = false
+    }
+    return exist
+}
+
+def hasUnstagedChanges(gitRepo) {
+    def hasUnstagedChanges = true
+    def gitStatusCmd = "git status".execute(null, gitRepo)
+    gitStatusCmd.waitFor()
+    def gitStatusOutput = gitStatusCmd.text
+    if (gitStatusOutput.contains("nothing to commit, working directory clean")) {
+        hasUnstagedChanges = false
+    }
+    return hasUnstagedChanges
+}
+
+def generateCheckstyleReport(localGitRepo, branch, checkstyleCfg, projectsToTestOn, destDir) {
+    println "Installing Checkstyle artifact ($branch) into local Maven repository ..."
+    executeCmd("git checkout $branch", localGitRepo)
+    executeCmd("mvn -Pno-validations clean install", localGitRepo)
+    executeCmd("groovy launch.groovy $projectsToTestOn $checkstyleCfg")
+    println "Moving Checkstyle report into $destDir ..."
+    moveDir("reports", destDir)
```

Comment:
Just confirming since you move directories around, did you confirm refs still point to valid location in final report for diff-tool?
For my tool I had to modify them.
https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff.sh#L271